### ISSUE TITLE
Correlated conditions are sometimes misapplied when two primary events have the same start date

### ISSUE THREAD

**Issue by xitology** [OHDSI/circe-be#163]
The issue could be reproduced on the cohort 436073000/436073003.json from PhenotypeLibrary on the synpuf 2.3m dataset under AWS Redshift.

I deployed the synpuf dataset on Redshift using https://ohdsi-sample-data.s3.amazonaws.com/CDM5.3.0_DDL_Redshift_optimized.sql and https://ohdsi-sample-data.s3.amazonaws.com/CMSDESynPUF23m.sql. Then I used circe-1.9.4.jar to generate a SQL file from the cohort definition (https://gist.github.com/xitology/615a745a195af1a5e99c76547313f5ed#file-436073003-redshift-sql). Running this file generates a cohort with 883603 entries.
```
$ psql -d synpuf2m -f 436073003.redshift.sql
[...]
INSERT 0 883603
[...]
```
When I compare the result with the cohort I generated independently from the same definition, I see a number of discrepancies:
```
3173×4 DataFrame
  Row │ subject_id  start_date   end_date     definition_id
      │ Int32?      Dates.Date?  Dates.Date?  Int32?
──────┼─────────────────────────────────────────────────────
    1 │       4343  2008-03-11   2008-03-11               1
    2 │       4343  2008-03-11   2008-03-17               0
    3 │       4706  2009-12-18   2009-12-18               1
    4 │       4706  2009-12-18   2010-01-07               0
    5 │       6354  2008-09-26   2008-09-26               1
    6 │       6354  2008-09-26   2008-09-27               0
    7 │       6650  2009-02-25   2009-02-25               1
    8 │       6650  2009-02-25   2009-03-17               0
    9 │       8293  2008-04-18   2008-04-18               1
   10 │       8293  2008-04-18   2008-05-08               0
[...]
 3170 │    2326058  2009-07-04   2009-07-04               1
 3171 │    2326058  2009-07-04   2009-07-22               0
 3172 │    2326058  2010-11-14   2010-11-14               1
 3173 │    2326058  2010-11-14   2010-11-28               0
```
Here, `definition_id = 0` refers to the cohort generated by circe.

It is easy to verify that the cohort generated by circe is indeed incorrect. This is because the cohort definition implies that `cohort_start_date` must be equal to `cohort_end_date`, which is enforced by the condition `A.START_DATE <= DATEADD(day,CAST(0 as int),P.START_DATE) AND A.START_DATE >= DATEADD(day,CAST(0 as int),P.END_DATE)` on the correlated event: https://gist.github.com/xitology/615a745a195af1a5e99c76547313f5ed#file-436073003-redshift-sql-L116.
However the generated cohort contains a number of entries that violate this condition:
```
$ psql -d synpuf2m -c 'select count(*) from cohort where cohort_definition_id = 0 and cohort_start_date < cohort_end_date'
 count
-------
  1564
(1 row)
```

I believe the issue is caused by the CTE clause that defines `primary_event`, specifically, the non-deterministic definition of `event_id` column in https://gist.github.com/xitology/615a745a195af1a5e99c76547313f5ed#file-436073003-redshift-sql-L55:
```
ROW_NUMBER() OVER (PARTITION BY E.person_id  ORDER BY E.sort_date  ASC ) ordinal,
```
This expression generates the value of the column `primary_event.event_id`. Notice that the `ORDER BY` clause does not enforce the total order on the rows.

What must be happening here is that Redshift does not materialize `primary_events`. Instead, it evaluates the definition afresh on every occurrence of `FROM primary_events`. As a result, if there are two primary events with the same `start_date`, they may have their `event_id` exchanged.

I confirmed this by making the `ORDER BY` clause deterministic with
```
ROW_NUMBER() OVER (PARTITION BY E.person_id  ORDER BY E.sort_date  ASC, E.event_id ) ordinal,
```
and running the query again (https://gist.github.com/xitology/615a745a195af1a5e99c76547313f5ed#file-fixed-436073003-redshift-sql):
```
$ psql -d synpuf2m -f fixed-436073003.redshift.sql
[...]
DELETE 883603
INSERT 0 883648
[...]

$ psql -d synpuf2m -c 'select count(*) from cohort where cohort_definition_id = 0 and cohort_start_date < cohort_end_date'
 count
-------
     0
(1 row)
```

**Comment by chrisknoll** [comment]
Thank you for the report.  Is the issue that in one subquery context (ie: within the correlated queries) you get one `row_number()` yielded but withint he exact same query exeuction, a different part of the query (in a different subquery) the row_number() will yield a different `row_number()` for the same set of records?  I think I understand where you are coming from, and so, yes, I think that if  that is the case, then we should put some sort of 'tie-break' on the queries (such as event_id) to ensure relative ordering is maintained across the different subqueries.

**Comment by xitology** [comment]
> Thank you for the report. Is the issue that in one subquery context (ie: within the correlated queries) you get one `row_number()` yielded but withint he exact same query exeuction, a different part of the query (in a different subquery) the row_number() will yield a different `row_number()` for the same set of records? I think I understand where you are coming from, and so, yes, I think that if that is the case, then we should put some sort of 'tie-break' on the queries (such as event_id) to ensure relative ordering is maintained across the different subqueries.

Yes, I believe this is the case. At least it explains the results I see. It's particularly surprising because `primary_events.event_id` is defined only once in the query, in the CTE clause. However, it must be getting potentially different values on each `FROM primary_events` invocation.

Perhaps, the easiest way to fix it is to explicitly materialize `primary_events` using a temporary table.

**Comment by chrisknoll** [comment]
> Perhaps, the easiest way to fix it is to explicitly materialize primary_events using a temporary table.

I've spent a good deal of time looking at temp tables vs. CTEs, and I've found that in some environments, the CTEs are optimized by using the CTE as an expression that can be applied to other parts of the query to push down predicates and lead to a more optimized query plan.  MS SqlServer does this, and even [PostgreSQL 14](https://www.postgresql.org/docs/14/queries-with.html) is going to support the notion of materialized and non-materialized CTEs.  So, I decided that leave it as CTE and if it's materialized, that's the same as creating a temp table, and if not, then we get the benefits of CTE optimization.  Going always with a temp table removes that option.

But, to your point, it's important that the CTEs result in a consistent state so I think the additional sort predicates could be warranted here, but I'm also worried about potential performance impact.  Could I ask you this: in your redshift environment: how are your tables partitioned/distributed (should be on person_id)? And do you apply any sort keys on the tables? (should be by person, date, and record id.  I'm wondering if we have to have explicit table ordering here?

**Comment by xitology** [comment]
For Redshift deployment, I used the schema and the data deployment file which I found in the https://github.com/OHDSI/OHDSIonAWS project, although I didn't use OHDSIonAWS directly, and instead deployed them manually. Here they are:
- https://ohdsi-sample-data.s3.amazonaws.com/CDM5.3.0_DDL_Redshift_optimized.sql
- https://ohdsi-sample-data.s3.amazonaws.com/CMSDESynPUF23m.sql

In `CDM5.3.0_DDL_Redshift_optimized.sql`, the event tables are decorated with `DISTKEY(person_id) compound sortkey(person_id, <event_type>_concept_id)`, so `sortkey` doesn't include the date or the event id.

I use a single-node dc2.large Redshift instance.

Is using the original event id instead of the synthetic id generated with row_number() an option? I'm not sure if primary events are guaranteed to be unique.
